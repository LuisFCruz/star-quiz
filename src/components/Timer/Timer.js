import { withStyles } from '@material-ui/core';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { finishGame } from '../../actions';

const styles = {
  clock: {
    fontWeight: '700',
    fontSize: '30px',
  },
};

export class Timer extends Component {
  static propTypes = {
    duration: PropTypes.number.isRequired,
    classes: PropTypes.object.isRequired,
    start: PropTypes.bool,
    finishGame: PropTypes.func
  }

  constructor(){
    super();
    
    this.state = {
      stopwatch: '',
      secondsRemaining: 0
    };

    this.interval = null;
  }

  componentDidMount = () => {
    const { duration = 0, start } = this.props;
    this.setSecondsRemaining(duration);

    if (start) { this.startCountDown(); }
  }

  componentDidUpdate = () => {
    const { start } = this.props;
    if (start) { this.startCountDown(); }
  }

  componentWillUnmount() {
    if (this.interval) { clearInterval(this.interval); }
  }

  startCountDown = () => {
    if (this.interval) { return; }

    this.interval = setInterval(() =>
      this.setSecondsRemaining(this.state.secondsRemaining - 1),
      1000
    );
  }

  secondsToMinutes = (seconds) => {
    const oneSecond = 60;
    const minute = Math.floor(seconds / oneSecond);
    const second = (seconds % oneSecond);
    const minutesString = minute < 10 ? `0${minute}` : `${minute}`;
    const secondsString = second < 10 ? `0${second}` : `${second}`;
    return `${minutesString}:${secondsString}`;
  }

  setSecondsRemaining(secondsRemaining) {
    if (secondsRemaining <= 0) {
      clearInterval(this.interval);
      this.props.finishGame(true);
    }

    if (secondsRemaining >= 0) {
      const stopwatch = this.secondsToMinutes(secondsRemaining);
      this.setState({ secondsRemaining, stopwatch });
    }
  }

  render() {
    const { classes } = this.props;

    return (
      <div className={classes.clock}>
        {this.state.stopwatch}
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  const { start } = state;
  return { start };
};

export default connect(mapStateToProps, { finishGame })(withStyles(styles)(Timer));
