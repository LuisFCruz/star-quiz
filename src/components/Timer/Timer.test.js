import React from 'react';
import { shallow } from 'enzyme';

import { Timer } from './Timer';

describe('<Timer />', () => {
  const props = {
    classes: {},
    duration: 120,
    start: false,
  };
  const wrapper = shallow(<Timer {...props} />);

  test('smoke test', () => {
    expect(Timer).toBeDefined();
    expect(wrapper.instance()).toBeInstanceOf(Timer);
  });

  test('render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should transform seconds to minutes ', () => {
    expect(wrapper.instance().secondsToMinutes).toBeDefined();
    const minutes = wrapper.instance().secondsToMinutes(60);
    expect(minutes).toBe('01:00');
  }); 

  test('should have set timer', () => {
    const seconds = 115;
    const component =  wrapper.instance();

    expect(component.setSecondsRemaining).toBeDefined();

    component.setSecondsRemaining(seconds);

    expect(wrapper.state('stopwatch')).toBe('01:55');
    expect(wrapper.state('secondsRemaining')).toBe(seconds);
  });

  test('should have call action finishGame()', () => {
    //Implementar quando estudar testes de ações do redux
  });
});
