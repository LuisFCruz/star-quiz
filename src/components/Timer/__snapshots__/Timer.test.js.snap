// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timer /> render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Timer
    classes={Object {}}
    duration={120}
    start={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "02:00",
      "className": undefined,
    },
    "ref": null,
    "rendered": "02:00",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "02:00",
        "className": undefined,
      },
      "ref": null,
      "rendered": "02:00",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
